libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt

```diff
@@ -270,34 +340,52 @@ class MapboxRouteLineView(var options: MapboxRouteLineOptions) {
         style: Style,
         clearRouteLineValue: Expected<RouteLineError, RouteLineClearValue>
     ) {
-        initializeLayers(style, options)
+        MapboxRouteLineUtils.initializeLayers(style, options)
         jobControl.scope.launch(Dispatchers.Main) {
             mutex.withLock {
-                clearRouteLineValue.onValue {
-                    updateSource(
-                        style,
-                        RouteLayerConstants.PRIMARY_ROUTE_SOURCE_ID,
-                        it.primaryRouteSource
-                    )
-                    if (it.alternativeRouteSourceSources.isNotEmpty()) {
-                        updateSource(
-                            style,
-                            RouteLayerConstants.ALTERNATIVE_ROUTE1_SOURCE_ID,
-                            it.alternativeRouteSourceSources.first()
-                        )
+                clearRouteLineValue.onValue { value ->
+                    val primarySourceKey = getSourceKeyForPrimaryRoute(style).fold(
+                        { routeSourceKey ->
+                            routeSourceKey
+                        }, { error ->
+                        logE(TAG, error.message)
+                        null
                     }
-                    if (it.alternativeRouteSourceSources.size > 1) {
+                    )?.also {
                         updateSource(
                             style,
-                            RouteLayerConstants.ALTERNATIVE_ROUTE2_SOURCE_ID,
-                            it.alternativeRouteSourceSources[1]
+                            it.sourceId,
+                            value.primaryRouteSource
                         )
+                        featureToSourceMap[it] =
+                            RouteLineFeatureId(
+                                value.primaryRouteSource.features()?.firstOrNull()?.id()
+                            )
                     }
+                    sourceLayerMap.keys.filter { it != primarySourceKey }
+                        .forEachIndexed { index, routeLineSourceKey ->
+                            if (index < value.alternativeRouteSourceSources.size) {
```

Comment:
The incoming `RouteLineClearValue` has a feature collection for the primary route and a list of feature collections for the alternatives.  The code is checking which source is hosting the primary route and which sources host the alternative routes.  I don't think this method should take for granted the incoming sources are all the same.  The `MapboxRouteLineApi` is making the decisions regarding the feature collections that are created, the `MapboxRouteLIneView` shouldn't make decisions that might be contradictory. It should render the data it receives.  The view class has always worked this way.

Reply:
It looks counter intuitive to me that a `ClearRouteLine` function would not clear all the lines, the `RouteLineClearValue` seems overloaded with data without a clear purpose. Just because something has always worked a certain way doesn't mean that we can't challenge and improve it.

Thanks for context though, it's not something we can change right now anyway, but could be a candidate for future refactors.